src/trusted/service_runtime/nacl_syscall_common.c

```diff
@@ -765,27 +762,23 @@ int32_t NaClSysOpen(struct NaClAppThread  *natp,
   
   if (retval < 0) {
     NaClLog(1, "Open returned error %d\n", retval);
+    free(hd);
     return retval;
   }
-  
-  fd_retval = NaClSetAvail(nap, ((struct NaClDesc *) NaClDescIoDescMake(hd)));
-  NaClLog(1, "Entered into open file table at %d\n", retval);
 
-
-cleanup:
   /*
-    Now translate the real fds to virtual fds and return them to the cage 
-    Get next available, and set cagetable there to nacl retval
+    Create NaclDesc and allocate a userfd.
   */
-  
-  retval = NextFd(nap->cage_id);
-  if(hd) {
-    hd->userfd = retval;
+
+  userfd = AllocNextFd(nap, hd);
+  if (userfd < 0) {
+    free(hd);
+    return -NACL_ABI_ENFILE;
```

Comment:
Yeah this is what I was talking about at the meeting about what's complicated. The issue here is that it will leave an orphaned file descriptor within rustposix which is why atomically reserving/assigning at the end is a minefield. I think you can probably just do a lind_close if need be there but I don't particularly like that and more thought should be put into potentially handling cases like this.